server/paymentProviders/stripe/webhook.ts

```diff
@@ -167,56 +168,98 @@ export const paymentIntentSucceeded = async (event: Stripe.Event) => {
+          previousPaymentIntents: [...(order.data.previousPaymentIntents ?? []), paymentIntent],
+        },
+      },
+      {
+        transaction: sequelizeTransaction,
+      },
     );
-    return;
-  }
 
-  await createOrUpdateOrderStripePaymentMethod(order, stripeAccount, paymentIntent);
+    await sequelizeTransaction.commit();
+    sequelizeTransaction = null;
 
-  const transaction = await createChargeTransactions(charge, { order });
-  const sideEffects: Promise<unknown>[] = [
-    order.update({
-      status: !order.SubscriptionId ? OrderStatuses.PAID : OrderStatuses.ACTIVE,
-      processedAt: new Date(),
-      data: {
-        ...omit(order.data, 'paymentIntent'),
-        previousPaymentIntents: [...(order.data.previousPaymentIntents ?? []), paymentIntent],
-      },
-    }),
-    order.getOrCreateMembers(),
-  ];
+    // If charge was already processed, ignore event. (Potential edge-case: if the webhook is called while processing a 3DS validation)
+    const existingChargeTransaction = await models.Transaction.findOne({
+      where: { data: { charge: { id: charge.id } } },
+    });
+    if (existingChargeTransaction) {
+      logger.info(
+        `Stripe Webhook: ${existingChargeTransaction.OrderId} already processed charge ${charge.id}, ignoring event ${event.id}`,
+      );
+      return;
+    }
 
-  // after successful first payment of a recurring subscription where the payment confirmation is async
-  // and the subscription is managed by us.
-  if (order.interval && !order.SubscriptionId) {
-    sideEffects.push(createSubscription(order));
-  }
+    await createOrUpdateOrderStripePaymentMethod(order, stripeAccount, paymentIntent);
+
+    const transaction = await createChargeTransactions(charge, { order });
+    const sideEffects: Promise<unknown>[] = [
+      order.update({
+        status: !order.SubscriptionId ? OrderStatuses.PAID : OrderStatuses.ACTIVE,
+        processedAt: new Date(),
+        data: {
+          ...omit(order.data, 'paymentIntent'),
+          previousPaymentIntents: [...(order.data.previousPaymentIntents ?? []), paymentIntent],
+        },
+      }),
+      order.getOrCreateMembers(),
+    ];
+
+    // after successful first payment of a recurring subscription where the payment confirmation is async
+    // and the subscription is managed by us.
+    if (order.interval && !order.SubscriptionId) {
+      sideEffects.push(createSubscription(order));
+    }
 
-  sendEmailNotifications(order, transaction);
-  await Promise.all(sideEffects);
+    sendEmailNotifications(order, transaction);
+    await Promise.all(sideEffects);
+  } catch (err) {
+    logger.error(
+      `Stripe Webhook: paymentIntentSuceeded errored on paymentIntentId ${paymentIntent.id}, OrderId: ${order?.id}: ${err}`,
+      err,
+    );
+
+    if (order) {
+      await order.update({
+        data: previousOrderData,
+      });
+    }
+
+    if (sequelizeTransaction) {
+      sequelizeTransaction.rollback();
+    }
```

Comment:
Two things here:
1. Should we rather rollback before we update?
2. It looks like we may end up calling `rollback` on an already `commit`ed transaction (e.g. if `createOrUpdateOrderStripePaymentMethod` fails). What would be the effect of that?

Reply:
1. You're right, its best to put that rollback first to avoid deadlocking.
2. `sequelizeTransaction` is set to null right after its committed so this would be skipped.